x/cronos/keeper/evmhandlers/cancel_send_to_evm_chain.go

```diff
@@ -42,49 +42,57 @@ func init() {
 	)
 }
 
-// CancelSendToChainHandler handles `__CronosCancelSendToChain` log
-type CancelSendToChainHandler struct {
+// CancelSendToEvmChainHandler handles `__CronosCancelSendToEvmChain` log
+type CancelSendToEvmChainHandler struct {
 	gravitySrv    gravitytypes.MsgServer
 	cronosKeeper  cronoskeeper.Keeper
 	gravityKeeper types.GravityKeeper
 }
 
-func NewCancelSendToChainHandler(
+func NewCancelSendToEvmChainHandler(
 	gravitySrv gravitytypes.MsgServer,
 	cronosKeeper cronoskeeper.Keeper,
 	gravityKeeper types.GravityKeeper,
-) *CancelSendToChainHandler {
-	return &CancelSendToChainHandler{
+) *CancelSendToEvmChainHandler {
+	return &CancelSendToEvmChainHandler{
 		gravitySrv:    gravitySrv,
 		cronosKeeper:  cronosKeeper,
 		gravityKeeper: gravityKeeper,
 	}
 }
 
-func (h CancelSendToChainHandler) EventID() common.Hash {
-	return CancelSendToChainEvent.ID
+func (h CancelSendToEvmChainHandler) EventID() common.Hash {
+	return CancelSendToEvmChainEvent.ID
 }
 
 // Handle `__CronosCancelSendToChain` log only if gravity is activated.
-func (h CancelSendToChainHandler) Handle(
+func (h CancelSendToEvmChainHandler) Handle(
 	ctx sdk.Context,
 	_ common.Address,
+	topics []common.Hash,
 	data []byte,
 	_ func(contractAddress common.Address, logSig common.Hash, logData []byte),
 ) error {
 	if h.gravitySrv == nil {
-		return fmt.Errorf("native action %s is not implemented", CancelSendToChainEventName)
+		return fmt.Errorf("native action %s is not implemented", CancelSendToEvmChainEventName)
+	}
+
+	if len(topics) != 2 {
+		// log and ignore
+		h.cronosKeeper.Logger(ctx).Info("log signature matches but wrong number of indexed events")
+		return nil
 	}
 
-	unpacked, err := CancelSendToChainEvent.Inputs.Unpack(data)
+	unpacked, err := CancelSendToEvmChainEvent.Inputs.Unpack(data)
 	if err != nil {
 		// log and ignore
 		h.cronosKeeper.Logger(ctx).Info("log signature matches but failed to decode")
 		return nil
 	}
 
-	senderCosmosAddr := sdk.AccAddress(unpacked[0].(common.Address).Bytes())
-	id := sdk.NewIntFromBigInt(unpacked[1].(*big.Int))
+	// needs to crope the extra bytes in the topic and cast to a cosmos address
+	senderCosmosAddr := sdk.AccAddress(common.BytesToAddress(topics[1].Bytes()).Bytes())
```

Comment:
What's the difference between decoding from topic and event data itself?

Reply:
indexed events are stored in topic

non indexed events are stored in data

https://docs.soliditylang.org/en/v0.8.16/abi-spec.html#events

Reply:
I see

Reply:
if the event log emitted by the contract has the right signature and accurate data(non indexed params), but indexed wrong params(either wrong `name` or wrong `type`), these indexed params would still be cast to a cosmos address and lead to unexpected result?

Reply:
type in the topics are always hash ( [32]byte)

we are already checking the length prior to ensure the number of arguments is correct but beside that, there are little check we could do (name and type does not matter)

I'm not sure also in which case such scenario could happen.

Reply:
I see. the params types should be guaranteed  by the signature, and name does not matter.